corehq/toggles/__init__.py

```diff
@@ -2698,7 +2698,7 @@ def domain_has_privilege_from_toggle(privilege_slug, domain):
 CUSTOM_DOMAIN_BANNER_ALERTS = StaticToggle(
     slug='custom_domain_banners',
     label='Allow projects to add banners for their users on HQ',
-    tag=TAG_CUSTOM,
+    tag=TAG_DEPRECATED,
```

Comment:
This was changed from a `FrozenPrivilegeToggle` back to a `StaticToggle`. Would the migrations for giving this privilege to relevant domains still work? Also, do all the privilege checks still work if this is a static toggle?

Reply:
> Would the migrations for giving this privilege to relevant domains still work?

The migration is meant to just add the new privilege so there should be no change there. It is independent of the toggle. So, in case of supporting relevant domains during GA, the new privilege isn't passed over to relevant domains but the access is handled in code. Shared useful code links about that later in the comment. Let me know if anything seems off for you.

> Also, do all the privilege checks still work if this is a static toggle?

Privilege checks only fallback to frozen toggles and not static toggles when needed. So after this change, the toggle is simply archived (deprecated) and is not referred to anymore.

Some more useful context:

The logic that determines if the privilege falls back to a feature flag, is handled in code. It's in the function that I replaced, `requires_privilege_with_fallback` and additionally in the function `request_has_privilege`. Both of these eventually checks for only frozen type toggles through `domain_has_privilege_from_toggle`

https://github.com/dimagi/commcare-hq/blob/f134cf08a19f770123c37edcf853c135c5e17d19/corehq/toggles/__init__.py#L2543-L2545

and we get that collection from

https://github.com/dimagi/commcare-hq/blob/f134cf08a19f770123c37edcf853c135c5e17d19/corehq/toggles/__init__.py#L2536-L2539

`requires_privilege_with_fallback` was replaced with `requires_privilege_raise404` so it does not check for toggle anymore. And `request_has_privilege` checks for frozen toggles only, so the deprecated feature flag isn't referred to anymore.

Reply:
Ah, thanks for the thorough explanation! I mistakenly thought the privilege was dependent on a `FrozenPrivilegeToggle`  to work correctly. Given that we aren't referring to the frozen privilege at all anymore, all good from my side!

Reply:
Not blocking, but I'm curious if there was a specific reason for removing the feature flag entirely instead of converting it into a frozen privilege toggle? This was the norm for GA'ing features flags in the past, so I'd just like to understand the different approach taken here.

Reply:
Hey @zandre-eng 

Yup, that is different from previous GA processes.
It is because we didn't have any projects that got access to the feature through feature flag and didn't have the necessary subscription to have access to the feature once it would become GA, so there was no need to retain the feature flag.
More in the comment [here](https://dimagi.atlassian.net/browse/SC-3198?focusedCommentId=308543)